lms/confirm-touch-chains

anathomical opened the pull request:
## WHAT
Update the way we bubble-up touches with custom callbacks
## WHY
To avoid some performance issues, we've had to write some custom callbacks to simulate `touch` behavior on a couple of models.  One of those custom callbacks wasn't being called in all situations.  The hope is that this fixes our cache invalidation process so that we can turn caching back on if this code doesn't cause performance issues of its own.
## HOW
1. Add an `after_touch` hook to `ClassroomUnit` as this is a likely cause of our failure to invalidate caches (`ActivitySession` calls `touch` on `ClassroomUnit`, but that wasn't triggering the `touch_classroom_without_callbacks` function, so we weren't managing to update the `Classroom.updated_at` values which are the key value in our current caching regime).
2. Instead of using `after_touch :save` in `Unit`, call our custom touch function directly.  `save` may trigger other callbacks which we don't want to pay the overhead for, so this is a little bit safer (even though the original way we were doing it didn't seem to be causing issues at the moment).
3. Add tests to our models to ensure that our touch callbacks trigger on `save` AND `touch`

### Notion Card Links
https://www.notion.so/quill/Cache-invalidation-fix-a5902d4aa1f04598926a635949ba098b

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Deploying now
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | N/A

anathomical committed:
Update the way we bubble-up touches with custom callbacks

To avoid some performance issues, we've had to write some custom callbacks to simulate `touch` behavior on a couple of models.  One of those custom callbacks wasn't being called in all situations.  The hope is that this fixes our cache invalidation process so that we can turn caching back on if this code doesn't cause performance issues.

anathomical committed:
Revert a change to this test that was made to handle an after_touch

Now that we've tweaked the after_touch not to trigger after_save hooks, this change is no longer needed and we can use the original version of this test

dandrabik commented on services/QuillLMS/app/models/classroom_unit.rb:
```diff
@@ -38,6 +38,7 @@ class ClassroomUnit < ApplicationRecord
   validates :unit, uniqueness: { scope: :classroom }
   before_save :check_for_assign_on_join_and_update_students_array_if_true
   after_save  :hide_appropriate_activity_sessions
+  after_touch :touch_classroom_without_callbacks
```
If we're doing `after_save` and `after_touch`, I think we could just do one `after_commit` instead (might be worth confirming that).

anathomical replied:
That seems reasonable.  Let me do some testing on that front to confirm.

anathomical replied:
Testing and reading shows that this should work as intended, so I'm making the suggested update.

dandrabik reviewed (approved):
I left a comment that there might be a way to make this shorter, but otherwise, looks good to me.

anathomical committed:
Consolidte after_touch + after_save hooks into a single after_commit

anathomical committed:
Check to make sure that underlying classrooms exist before touching

Classroom has a `cascade: :destroy` chain that, among other things, destroys related classroom units.  A destruction event causes a commit, which triggers our after_commit touch function.  However, in this scenario, the classroom that we specify to be touched has been destroyed, so obviously can't be updated.  This guard clause make sure we don't cause an error in these cases.
